### DISCUSSION TITLE
Invalid (or missing) WebSocket implementation with SSR (server-side rendering)

### DISCUSSION THREAD

**Discussion by furkandmrblk** [enisdenjo/graphql-ws#215] [Q&A] [answered]
Hi, I am using Apollo-Express for my backend, so I implemented the given example from this repo. For the frontend, I also used the given link & created a splitLink for my httpLink & wsLink. My issue is that my frontend is giving me the following error: ```Invalid WebSocket implementation provided```
I do not have any errors in my network in the chrome browser nor do I have any errors in my backend. 

<img width="1680" alt="Screenshot 2021-07-21 at 01 15 03" src="https://user-images.githubusercontent.com/74838441/126402227-87b6b1bc-8330-441e-b20d-133dc08ec651.png">

Here is a snippet of my backend code that is affiliated with graphql-ws: 
```
const server = app.listen(port, () => {
    const wsServer = new ws.Server({
      server: server,
      path: '/graphql',
    });

    useServer(
      {
        schema: typeDefs as any,
        roots,
        context: ({ extra }) => ({ req: extra.request }),
        // execute,
        // subscribe,
      },
      wsServer
    );

    console.log(`Server is ready at http://localhost:${port}`);
  });
```

And here is the frontEnd code: 
```
import {
  ApolloClient,
  split,
  createHttpLink,
  from,
  InMemoryCache,
  NormalizedCacheObject,
} from '@apollo/client';

import { TokenRefreshLink } from 'apollo-link-token-refresh';

import {
  accessTokenExpired,
  fetchAccessToken,
  getAccessToken,
  setAccessToken,
} from '../helpers/Tokens';

import { setContext } from '@apollo/client/link/context';
import { onError } from '@apollo/client/link/error';

// import { WebSocketLink } from '@apollo/client/link/ws';

import { getMainDefinition } from '@apollo/client/utilities';
import { WebSocketLink } from './WebSocket';
import WebSocket from 'ws';

const wsLink = new WebSocketLink({
  url: 'ws://localhost:4000/graphql',
  connectionParams: () => {
    const accessToken = getAccessToken().valueOf();

    if (!accessToken) return {};

    return {
      Authorization: accessToken ? `Bearer ${accessToken}` : '',
    };
  },
  webSocketImpl: WebSocket,
});

const httpLink = createHttpLink({
  uri: 'http://localhost:4000/graphql',
  credentials: 'include',
});

const splitLink = split(
  ({ query }) => {
    const definition = getMainDefinition(query);
    return (
      definition.kind === 'OperationDefinition' &&
      definition.operation === 'subscription'
    );
  },
  wsLink,
  httpLink
);

const refreshLink: any = new TokenRefreshLink({
  accessTokenField: 'accessToken',
  isTokenValidOrUndefined: () => accessTokenExpired(),
  fetchAccessToken: () => fetchAccessToken(),
  handleFetch: (accessToken: string) => setAccessToken(accessToken),
  handleResponse: (operation, accessTokenField) => (response) => {
    if (!response) return { accessToken: null };

    return { accessToken: response.data?.updateTokens?.accessToken };
  },
  handleError: (err: Error) => {
    console.log('An error occurred');
    console.log(err);
  },
});

const authLink = setContext((_, { headers }) => {
  const accessToken = getAccessToken().valueOf();

  return {
    headers: {
      ...headers,
      authorization: accessToken ? `Bearer ${accessToken}` : '',
    },
  };
});

const errorLink = onError(({ graphQLErrors, networkError }) => {
  if (graphQLErrors)
    graphQLErrors.map(({ message, locations, path }) =>
      console.log(
        `[GraphQL error]: Message: ${message}, Location: ${locations}, Path: ${path}`
      )
    );
  if (networkError) console.log(`[Network error]: ${networkError}`);
});

export const client: ApolloClient<NormalizedCacheObject> = new ApolloClient({
  ssrMode: true,
  cache: new InMemoryCache(),
  link: from([refreshLink, authLink, errorLink, splitLink]),
  defaultOptions: {
    watchQuery: {
      fetchPolicy: 'cache-and-network',
    },
  },
});
```

The WebSocketLink:
```
import { Client, createClient, ClientOptions } from 'graphql-ws';
import { print, GraphQLError } from 'graphql';

import {
  ApolloLink,
  FetchResult,
  Observable,
  Operation,
} from '@apollo/client/core';

export class WebSocketLink extends ApolloLink {
  private client: Client;

  constructor(options: ClientOptions) {
    super();
    this.client = createClient(options); <==================== The error is pointing at this place 
  }

  public request(operation: Operation): Observable<FetchResult> {
    return new Observable((sink) => {
      return this.client.subscribe<FetchResult>(
        { ...operation, query: print(operation.query) },
        {
          next: sink.next.bind(sink),
          complete: sink.complete.bind(sink),
          error: (err) => {
            if (err instanceof Error) {
              return sink.error(err);
            }

            if (err instanceof CloseEvent) {
              return sink.error(
                // reason will be available on clean closes
                new Error(
                  `Socket closed with event ${err.code} ${err.reason || ''}`
                )
              );
            }

            return sink.error(
              new Error(
                (err as GraphQLError[]).map(({ message }) => message).join(', ')
              )
            );
          },
        }
      );
    });
  }
}
```
Sorry for the inconvenience & thank you for taking your time to read this.

**Comment by enisdenjo** [comment]
Hey there, the error claims that the `webSocketImpl` you provided is invalid - it probably is. Are you sure you have `ws` installed on the frontend? What does the WebSocket from `import WebSocket from 'ws';` look like?

Nevertheless, if you're running the client in a browser, you dont even need `ws`. Browsers have their own WebSocket implementation which I would absolutely recommend using instead (`ws` module is just for backend/NodeJS environments).

```diff
const wsLink = new WebSocketLink({
  url: 'ws://localhost:4000/graphql',
  connectionParams: () => {
    const accessToken = getAccessToken().valueOf();

    if (!accessToken) return {};

    return {
      Authorization: accessToken ? `Bearer ${accessToken}` : '',
    };
  },
- webSocketImpl: WebSocket,
});
```

P.S. The ["Client usage with Apollo" recipe](https://github.com/enisdenjo/graphql-ws#apollo-client) does not use the `webSocketImpl`.

**Comment by furkandmrblk** [comment]
Hey, thank you for the quick answer!

Yeah I imported it correctly and I also tried it without the ```webSocketImpl: WebSocket``` but then I get the following error: 
```WebSocket Implementation missing``` with a status code of 500 for the localhost in the network tab.

<img width="1680" alt="Screenshot 2021-07-21 at 12 51 57" src="https://user-images.githubusercontent.com/74838441/126469832-513284ae-7f1a-4a1b-bf8c-d3cb3349fdce.png">

**Reply by enisdenjo** [reply]
Status code 500 can not be caused by `graphql-ws`. Are you sure your frontend is running in a browser? How are you running your app, whats your stack?

**Reply by enisdenjo** [reply]
My guess is that this is caused by server-side rendering, trying to create the client on the server before sending the rendered content to the client. I am no expert in SSR, but I recon there must be a way to flag parts available in browser contexts only.

**Reply by furkandmrblk** [reply]
I am running this in google chrome and I am using the new macbooks with the m1 chips. I am fairly new to coding so I don't know if it could be because of the m1 chip. I am using Next.js & Apollo for the frontend and Node.js with Apollo-Express for the backend. I will look into the SSR and give you an update. Thank you again for the quick response!

**Reply by enisdenjo** [reply]
@furkandmrblk maybe try running the client with a isomorphic W3C WebSocket implementation. Try this:
1. `yarn add websocket` or `npm i websocket`
2. Use `import { w3cwebsocket } from 'websocket';` as the implementation like this:
```js
import { w3cwebsocket } from 'websocket';

const wsLink = new WebSocketLink({
  url: 'ws://localhost:4000/graphql',
  connectionParams: () => {
    const accessToken = getAccessToken().valueOf();

    if (!accessToken) return {};

    return {
      Authorization: accessToken ? `Bearer ${accessToken}` : '',
    };
  },
  webSocketImpl: w3cwebsocket,
});
```

**Reply by enisdenjo** [reply]
M1 chips have nothing to do with it. [Next.js](https://nextjs.org/) is often used for server-side rendering React apps, which is why I think it is a SSR issue.

Please try my suggestion from above.

**Reply by praful-chandra** [reply]
I had faced the same issue with the SSR.
try to execute the WebSocketLink only when process.browser is true.

example

```
const wsLink = process.browser ? new WebSocketLink({
  url: "ws://localhost:4000/graphql",
  connectionParams: () => {
    return {
      authorization: `Bearer `,
    };
  }
}) : null
```

**Reply by furkandmrblk** [reply]
Thank you guys, the example above with the ```import { w3cwebsocket } from 'websocket' ``` worked! And I will also try out the solution from @praful-chandra. Thanks guys for all the help!

**Reply by enisdenjo** [reply]
If @praful-chandra example works, I'd recommend using that instead because it does not require additional modules and feels much cleaner.

@furkandmrblk please ping me if his example works. I'd like to aggregate the two suggestions in one answer.

**Reply by furkandmrblk** [reply]
@enisdenjo the example from @praful-chandra gives me another error at my splitLink. The package is working fine.
<img width="1680" alt="Screenshot 2021-07-21 at 14 09 11" src="https://user-images.githubusercontent.com/74838441/126479429-b0802d87-2be2-4691-a0fa-df03b8339420.png">

**Reply by furkandmrblk** [reply]
I will try to figure this out too. If I can work it out without the package I will give you an update.

**Reply by praful-chandra** [reply]
try this in split

```
const splitLink = process.browser ? split( //only create the split in the browser
  // split based on operation type
  ({ query }) => {
    const definition = getMainDefinition(query);
    return (
      definition.kind === 'OperationDefinition' &&
      definition.operation === 'subscription'
    );
  },
  wsLink ?? httpLink,
  httpLink,
) : httpLink;
```

**Reply by enisdenjo** [reply]
This is because you're using the `splitLink` and it is `null`. @praful-chandra example completely disables the `WebSocketLink` if it is not running in a browser environment.

Please refactor your code so that you dont use the WebSocketLink until `process.browser === true`. Like @praful-chandra showcased above.

**Reply by furkandmrblk** [reply]
Yeah it works, my solution isn't the best right now and I will try to make it clean. But I can assure you that his solution works

**Reply by furkandmrblk** [reply]
This is my work around for now:
```
const wsLink = process.browser
  ? new WebSocketLink({
      url: 'ws://localhost:4000/graphql',
      connectionParams: () => {
        const accessToken = getAccessToken().valueOf();

        return {
          Authorization: accessToken ? `Bearer ${accessToken}` : '',
        };
      },
    })
  : null;

const httpLink = createHttpLink({
  uri: 'http://localhost:4000/graphql',
  credentials: 'include',
});

const splitLink =
  wsLink !== null
    ? split(
        ({ query }) => {
          const definition = getMainDefinition(query);
          return (
            definition.kind === 'OperationDefinition' &&
            definition.operation === 'subscription'
          );
        },
        wsLink,
        httpLink
      )
    : createHttpLink({
        uri: 'http://localhost:4000/graphql',
        credentials: 'include',
      });

export const client: ApolloClient<NormalizedCacheObject> = new ApolloClient({
  // ssrMode: true,
  cache: new InMemoryCache(),
  link: from([refreshLink, authLink, errorLink, splitLink]),
  defaultOptions: {
    watchQuery: {
      fetchPolicy: 'cache-and-network',
    },
  },
});
```
Thank you guys again!

**Reply by enisdenjo** [reply]
Awesome! Glad you got it working.

**Comment by enisdenjo** [comment, marked as answer]
The issue here is that the client is being instantiated during server-side rendering on the server. Because the browser's WebSocket is not available there, this issue appears.

**Recommended** solution, as suggested by @praful-chandra in https://github.com/enisdenjo/graphql-ws/discussions/215#discussioncomment-1031984, is to skip over WebSocket parts when not in a browser environment.
```js
const wsLink = typeof window === 'undefined' ? null : new WebSocketLink({
  url: "ws://localhost:4000/graphql",
  connectionParams: () => {
    const accessToken = getAccessToken().valueOf();

    if (!accessToken) return {};

    return {
      Authorization: accessToken ? `Bearer ${accessToken}` : '',
    };
  },
});
```
_`process.browser` can be used for checking the environment too; but beware, it is deprecated: https://github.com/vercel/next.js/issues/5354#issuecomment-520305040_

---

An alternative, would be to use the [isomorphic W3CWebSocket class from `websocket`](https://github.com/theturtle32/WebSocket-Node#client-example-using-the-w3c-websocket-api).
```js
import { w3cwebsocket } from 'websocket';

const wsLink = new WebSocketLink({
  webSocketImpl: w3cwebsocket,
  url: 'ws://localhost:4000/graphql',
  connectionParams: () => {
    const accessToken = getAccessToken().valueOf();

    if (!accessToken) return {};

    return {
      Authorization: accessToken ? `Bearer ${accessToken}` : '',
    };
  },
});
```
